DQM/SiPixelHeterogeneous/plugins/SiPixelCompareRecHits.cc

```diff
@@ -0,0 +1,283 @@
+#include "DataFormats/SiPixelDetId/interface/PixelSubdetector.h"
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
+#include "DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsHost.h"
+#include "DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsSoA.h"
+#include "CUDADataFormats/TrackingRecHit/interface/TrackingRecHitSoAHost.h"
+#include "CUDADataFormats/TrackingRecHit/interface/TrackingRecHitsUtilities.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
+#include "Geometry/CommonTopologies/interface/PixelTopology.h"
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+
+template <typename T>
+class SiPixelCompareRecHits : public DQMEDAnalyzer {
+public:
+  using HitsAlpaka = TrackingRecHitHost<T>;
+  using HitsCUDA = TrackingRecHitSoAHost<T>;
+
+  explicit SiPixelCompareRecHits(const edm::ParameterSet&);
+  ~SiPixelCompareRecHits() override = default;
+  void dqmBeginRun(const edm::Run&, const edm::EventSetup&) override;
+  void bookHistograms(DQMStore::IBooker& ibooker, edm::Run const& iRun, edm::EventSetup const& iSetup) override;
+  void analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) override;
+  // TODO: remove analyzeSeparate function below once the CUDA code is removed from CMSSW; it was implemented
+  // to make the DQM compare modules work for every case: CUDA vs CUDA, Alpaka vs Alpaka and Alpaka vs CUDA;
+  // the content of analyzeSeparate can be copied to analyze once the CUDA code is removed
+  template <typename U, typename V>
+  void analyzeSeparate(U tokenRef, V tokenTar, const edm::Event& iEvent);
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  const edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> geomToken_;
+  const edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> topoToken_;
+  // CUDA and Alpaka tokens are implemented to make the DQM compare modules work for every case:
+  // CUDA vs CUDA, Alpaka vs Alpaka and Alpaka vs CUDA; CUDA tokens must be removed together with rest of CUDA code
+  const edm::EDGetTokenT<HitsCUDA> tokenSoAHitsReferenceCUDA_;  //these two are both on CPU but originally they have been
+  const edm::EDGetTokenT<HitsCUDA> tokenSoAHitsTargetCUDA_;     //produced on CPU or on GPU
+  const edm::EDGetTokenT<HitsAlpaka>
+      tokenSoAHitsReferenceAlpaka_;  //these two are both on Host but originally they have been
+  const edm::EDGetTokenT<HitsAlpaka> tokenSoAHitsTargetAlpaka_;  //produced on Host or on Device
```

Comment:
```suggestion
  // these two are both on CPU but originally they have been produced on CPU or on GPU
  const edm::EDGetTokenT<HitsCUDA> tokenSoAHitsReferenceCUDA_;
  const edm::EDGetTokenT<HitsCUDA> tokenSoAHitsTargetCUDA_;
  // these two are both on Host but originally they have been produced on Host or on Device
  const edm::EDGetTokenT<HitsSoA> tokenSoAHitsReference_;
  const edm::EDGetTokenT<HitsSoA> tokenSoAHitsTarget_;
```

See [this comment](https://github.com/cms-sw/cmssw/pull/43964#issuecomment-1966013353).

Reply:
I agree with this and will make the same changes into the tracks and vertices comparisons as well